File: scripts/updateVersions.js
const { EOL } = require("os");
const { env } = require("process");
const { replaceInFile } = require("replace-in-file");

const versionFilesOption = {
  files: ["robotcode/_version.py", "pyproject.toml"],
  from: /(^_*version_*\s*=\s*['"])([^'"]*)(['"])/gm,
  to: `\$1${env.npm_package_version || ""}\$3`,
};
replaceInFile(versionFilesOption, function (error, _results) {
  if (error) {
    console.error(error);
  }
});

const changelogOptions = {
  files: ["CHANGELOG.md"],
  from: /^(\#*\s*)(\[Unreleased\])$/gm,
  to: `\$1\$2${EOL}${EOL}\$1 ${env.npm_package_version || ""}`,
};
replaceInFile(changelogOptions, function (error, _results) {
  if (error) {
    console.error(error);
  }
});

File: scripts/update_versions.py
import os
import re
from pathlib import Path

from scripts.tools import get_version


def replace_in_file(filename: Path, pattern: "re.Pattern[str]", to: str) -> None:
    text = filename.read_text()
    new = pattern.sub(to, text)
    filename.write_text(new)


def main() -> None:

    version = get_version()

    preview = version.minor % 2 != 0

    for f in ["robotcode/__version__.py", "pyproject.toml"]:
        replace_in_file(
            Path(f),
            re.compile(r"""(^_*version_*\s*=\s*['"])([^'"]*)(['"])""", re.MULTILINE),
            rf"\g<1>{version or ''}{'.dev.0' if preview and version.prerelease is None else ''}\g<3>",
        )

    for f in ["CHANGELOG.md"]:
        replace_in_file(
            Path(f),
            re.compile(r"^(\#*\s*)(\[Unreleased\])$", re.MULTILINE),
            rf"\g<1>\g<2>{os.linesep}- none so far{os.linesep}\g<1> {version or ''}",
        )


if __name__ == "__main__":
    main()
